contrib/labstack/echo.v4/appsec.go

```diff
@@ -6,33 +6,55 @@
 package echo
 
 import (
+	"net/http"
+
 	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
-	"gopkg.in/DataDog/dd-trace-go.v1/internal/appsec/dyngo/instrumentation"
 	"gopkg.in/DataDog/dd-trace-go.v1/internal/appsec/dyngo/instrumentation/httpsec"
 
 	"github.com/labstack/echo/v4"
 )
 
-func useAppSec(c echo.Context, span tracer.Span) func() {
-	instrumentation.SetAppSecEnabledTags(span)
-
-	params := make(map[string]string)
-	for _, n := range c.ParamNames() {
-		params[n] = c.Param(n)
-	}
-
-	req := c.Request()
-	ipTags, clientIP := httpsec.ClientIPTags(req.Header, req.RemoteAddr)
-	instrumentation.SetStringTags(span, ipTags)
-
-	args := httpsec.MakeHandlerOperationArgs(req, clientIP, params)
-	ctx, op := httpsec.StartOperation(req.Context(), args)
-	c.SetRequest(req.WithContext(ctx))
-	return func() {
-		events := op.Finish(httpsec.HandlerOperationRes{Status: c.Response().Status})
-		if len(events) > 0 {
-			httpsec.SetSecurityEventTags(span, events, args.Headers, c.Response().Writer.Header())
+func withAppSec(next echo.HandlerFunc, span tracer.Span) echo.HandlerFunc {
+	return func(c echo.Context) error {
+		params := make(map[string]string)
+		for _, n := range c.ParamNames() {
+			params[n] = c.Param(n)
 		}
-		instrumentation.SetTags(span, op.Tags())
+		var err error
+		// Wrap the echo response to allow monitoring of the response status code in httpsec.WrapHandler()
+		srw := &statusResponseWriter{r: c.Response()}
+		handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			c.SetRequest(r)
+			err = next(c)
+			srw.status = c.Response().Status
+			if e, ok := err.(*echo.HTTPError); ok && e != nil {
+				srw.status = e.Code
+			}
+		})
+		httpsec.WrapHandler(handler, span, params).ServeHTTP(srw, c.Request())
+		return err
 	}
+
+}
+
+// statusResponseWriter wraps an echo response to allow tracking/retrieving its status code through a Status() method
+// without having to rely on the echo error handlers
+type statusResponseWriter struct {
+	r      *echo.Response
+	status int
+}
+
+// Status returns the status code of the response
+func (w *statusResponseWriter) Status() int { return w.status }
+
+// WriteHeader wraps the underlying echo response writer WriteHeader() call
+func (w *statusResponseWriter) WriteHeader(statusCode int) {
```

Comment:
Side note on suspicious request blocking: that's exactly where you will need to call the WAF with this status code in order to be able to with the appsec response writer instead. (So the status code address will have to become a "floating http operation address" like the body for example).